Improve performance by 10%

Confusion opened the pull request:
Inspired by a blog post [1] on the 'allocation_stats' gem I looked into
yard's memory use (using the 'memory_profiler' gem) and found some hotspots
that the attached patch removes. It reduces memory allocations by ~25%,
while increasing performance by ~10%, measured on the test_doc Rake task of
Yard itself.

[1] https://blog.presidentbeef.com/blog/2018/11/28/finding-ruby-performance-hotspots-via-allocation-stats/

# Description

Describe your pull request and problem statement here.

# Completed Tasks

* [x] I have read the [Contributing Guide][contrib].
* [x] The pull request is complete (implemented / written).
* [x] Git commits have been cleaned up (squash WIP / revert commits).
* [x] I wrote tests and ran `bundle exec rake` locally (if code is attached to PR).

[contrib]: https://github.com/lsegal/yard/blob/master/CONTRIBUTING.md

lsegal commented on lib/yard/code_objects/namespace_mapper.rb:
```diff
@@ -68,6 +73,14 @@ def separators_match
         NamespaceMapper.map_match
       end
 
+      def starts_with_separator_regex
```
These two methods have a possible caching bug-- if `NamespaceMapper.invalidate` is called, these will not be invalidated along with `map_match`.

Confusion commented:
I renamed the methods, moved them to RegistryResolver as private methods (but documented them anyway) and hooked into `invalidate` and `default_seperator` to invalidate the memoizations when necessary.

This of course creates additional coupling between the RegistryResolver and the NamespaceMapper through the 'super' calls (which at least will break on renames and changes in number of parameters, so chance of silently losing functionality seems slim), but I don't think there is any way to avoid that.

Performance gain is still ~10% (actually it seems more like 15-20%, but I do not sufficiently trust my measurement methodology to claim that much)

lsegal commented on lib/yard/registry_resolver.rb:
```diff
@@ -185,5 +185,35 @@ def collect_namespaces(object)
     end
+
+    # @see NamespaceMapper#invalidate
+    def invalidate
```
Unfortunately this won't trigger as `Namespacemapper.invalidate` is a module function. 

The good news is I think you can/should override `#clear_separators` instead, since NamespaceMapper's singleton stuff internal and only expected to be used _through_ / exposed by those instance methods.

```suggestion
    def clear_separators
```

(Need to change the docstring too)

Confusion replied:
:+1: 

I now realize my previous comment wasn't true though: if `clear_separators` is renamed, this won't fail: it just silently won't be called. Do you want me to add a spec for that?

lsegal replied:
Sure a spec would help.

Confusion replied:
I just noticed that `invalidate` is also called via `register_separator`, so `register_separator` won't result in the memoizations being cleared, which seems necessary. So we need to extend `invalidate` after all? 

What we could do is add a line
```
invalidate if respond_to?(:invalidate)
```
to `register_separator`, after `NamespaceMapper.invalidate` so users of NamespaceMapper can define an instance method :invalidate for additional resets. What do you think about that?

Confusion replied:
I opted to extend both #register_separator and #clear_separators

lsegal commented on lib/yard/registry_resolver.rb:
```diff
@@ -185,5 +185,35 @@ def collect_namespaces(object)
 
       nss
     end
+
```
You should probably mark all these methods below as private since they are not part of the public API.

```suggestion

    private

```

Confusion replied:
There's a `private` marker on line 93 that covers these methods, so that's already taken care of?

lsegal replied:
You're right about that, I missed it! You can ignore this one.
